### PR TITLE
rate: Merge record_trial_encode into update_state

### PR DESCRIPTION
**Pull request by barrbrain**
This eliminates some code duplication and aligns with [Theora rate control](https://git.xiph.org/?p=theora.git;a=blob;f=lib/rate.c;h=bf2b1396a17ea3c7d7ac87e8b78771f7874a4a8a;hb=HEAD#l732) design.

### CODE REVIEW

File: src/rate.rs [hunk 1]
```diff
@@ -786,54 +787,34 @@ impl RCState {
...
 
   pub fn needs_trial_encode(&self, fti: usize) -> bool {
       self.target_bitrate > 0 && self.nframes[fti] == 0
   }
-
-  pub fn record_trial_encode(
-    &mut self, bits: i64, fti: usize, log_target_q: i64
-  ) {
-    assert!(self.needs_trial_encode(fti));
-    assert!(bits > 0);
```
Comment (negge) [inline src/rate.rs]: These two assert!() statements did not make it.  Do we want to still check for them if trial is true?

File: src/rate.rs [hunk 2]
```diff
@@ -726,7 +726,8 @@ impl RCState {
   }
 
   pub fn update_state(
-    &mut self, bits: i64, fti: usize, log_target_q: i64, droppable: bool
+    &mut self, bits: i64, fti: usize, log_target_q: i64, trial: bool,
+    droppable: bool
   ) -> bool {
     let mut dropped = false;
     // Update rate control only if rate control is active.
```
Comment (negge) [inline src/rate.rs]: Is it possible that self.target_bitrate <= 0 when update_state() is called with trial == true?  If so the else side of this block will not execute and your trial encode will not be recorded.  Do we want an assert!() for this to catch changed behavior?

Comment (barrbrain) [reply src/rate.rs]: So, if `trial` implies `needs_trial_encode()` then this cannot be the case. Let's keep the asserts that were dropped.

File: src/rate.rs (hunk 1 above)
Comment (barrbrain) [reply src/rate.rs]: Yes, the above comment is related and where we should confirm the preconditions.

File: src/rate.rs [hunk 3]
```diff
@@ -747,7 +752,7 @@ impl RCState {
         // If this is the first example of the given frame type we've seen, we
         //  immediately replace the default scale factor guess with the
         //  estimate we just computed using the first frame.
-        if self.nframes[fti] <= 0 {
+        if trial || self.nframes[fti] <= 0 {
```
Comment (dwbuiten) [inline src/rate.rs]: Is there ever a time when only one of these is true? i.e. can they be merged?

Comment (barrbrain) [reply src/rate.rs]: At present, trial implies `self.nframes[fti] <= 0` but my intention was to be explicit about the dual use of this block.

Comment (dwbuiten) [reply src/rate.rs]: That's what I thought. No complaint from me if you prefer it.

**Comment by barrbrain** [comment]
> [02:45am KST] derf: barrbrain: Thanks for 1321. LGTM.

Tim also approves.

**Pull request merged**